### PR TITLE
Add a script/tool that allows introspection of our blobstore and cleaning up unreferenced blobs.

### PR DESCRIPTION
**Pull request by jameinel**
## Description of change

This adds 2 tools into the 'scripts/' directory. The first is "juju-list-blobstore", which starts by listing all models, and looking for references from the model to "resources". That includes Agent Binaries, Charm Archives, and Charm Resources (TBD, backups and images). It then walks the managedStoredResources collection to find other managed resources that have a bucketUUID that matches the modelUUID, but aren't actually referenced. After walking all models, it then iterates through:
 1. all managedStoredResources for any bucketUUIDs that aren't an active modelUUID
 2. all storedResources for anything that isn't referenced by *any* managed stored resource
 3. all blobstore.blobstore.files objects that aren't referenced by any storedResource
 4. all blobstore.blobstore.chunks objects that don't have an associated file.

This generates a report of what data is stored, how much of it is properly referenced, and how much is waste.

The other new tool is "juju-blobstore-cleanup". This defaults to '--dry-run', but with '--dry-run=false' will actively remove resources from the database. The algorithm we use is:

 1. Find managedStoredResources that don't belong to any active model. Since we know these can't actually be acted upon. When we remove these, we only remove the reference count, so if the resource is still referenced by an active model, it won't actually free up space. But that is fine, it still makes the database cleaner, and it means when an active model stops referencing it (say by upgrading a charm), then it will free up space.
 2. Find resources that are unreferenced. Since there are no references, the refcount is clearly wrong, but was just not cleaned up properly. Looking at the ManagedStorage code, the ManagedResource is removed before we remove the refcount. Which means that if the first step succeeded, but the second failed, we would end up with a dangling reference. These get pruned, along with their associated blobstore file.
 3. Find blobstore.blobstore.files that are unreferenced. I audited our codebase, and all references to GridFS were behind some sort of ManagedStorage. So we shouldn't have any blobstore.files that don't have an associated storedResources document. These are Removed. This also has the same issue in (2), where the reference is removed before the file, but if anything interrupts it, nothing will come by later and clean it up.
 4. Find blobstore.blobstore.chunks that don't refer to a files object. Likely this is either a partial upload, or possibly a partial delete. Either way, we know that we don't need them anymore.

I submitted https://bugs.launchpad.net/juju/+bug/1846310 to track the idea that we should be doing better at making sure we clean up even in the face of failures/errors. And https://bugs.launchpad.net/juju/+bug/1846311 to track cleaning up more resources that are no longer actually referenced.

## QA steps

Once this is built, you can use:
```
$ juju scp -m controller $GOPATH/bin/juju-list-blobstore 0:.
$ juju ssh -m controller 0
$$ sudo -i
$$ ~ubuntu/juju-list-blobstore
```

And it should print out the known blobstore usage like:
```
# ~ubuntu/juju-list-blobstore -h
Model: "controller"
  Referenced By Apps
  app referenced charm bytes: 0B
  total charm bytes: 0B
  No Referenced Resources

  total unreferenced model bytes: 0B
  total model bytes: 0B

Model: "default"
  Referenced By Apps
  - cs:postgresql-199: 5.0MB 30baf18b...
  - local:bionic/dummy-fat-sink-13: 105MB b7ccd9f8...
  - local:bionic/dummy-fat-sink-15: 105MB cde4cebe...
  app referenced charm bytes: 215MB
  Referenced By Units
  - local:bionic/dummy-fat-sink-14: 105MB 3f602fdc...
     - dummy-fat-sink/1
  unit referenced charm bytes: 105MB
  total charm bytes: 320MB
  Referenced Resources
  - application-postgresql/resources/wal-e: 105MB 725734d7...
  total resource bytes: 105MB

  total unreferenced model bytes: 0B
  total model bytes: 424MB

Model: "new"
  Referenced By Apps
  - cs:postgresql-199: 5.0MB 30baf18b...
  app referenced charm bytes: 5.0MB
  total charm bytes: 5.0MB
  No Referenced Resources

  total unreferenced model bytes: 0B
  total model bytes: 5.0MB

Model Summary
  app referenced charm bytes: 220MB
  unit referenced charm bytes: 105MB
  unreferenced charm bytes: 0B
  referenced resource bytes: 105MB
  unreferenced misc bytes: 0B
  total unreferenced bytes: 0B
  total model bytes: 430MB

Agent Binaries
  - 2.7-beta1.4: 49MB 66b90ef5...
  - 2.7-beta1.6: 49MB 3f2811da...
  referenced agent bytes: 98MB

Unreferenced Agent Binaries
  2.7-beta1.1-artful-amd64: 49MB 16 88f8d368...
  2.7-beta1.2-artful-amd64: 49MB 16 5edeea86...
  2.7-beta1.3-artful-amd64: 49MB 16 78bc724f...
  2.7-beta1.5-artful-amd64: 49MB 16 f6c2af0b...
  2.7-beta1.7-artful-amd64: 49MB 16 2d8aa5b6...
  unreferenced agent bytes: 246MB
total agent bytes: 344MB

No Unknown Resources

No Unknown Blobstore Files
  read 12 blobstore files

No Unknown Blobstore Chunks
  read 2951 blobstore chunks
```

You can also use `-v` to see details about who is referencing the various objects (what agents are running particular agent binaries, etc.)

For `juju-blobstore-cleanup` it defaults to --dry-run, so is safe to run. Using:
```
# ~ubuntu/juju-blobstore-cleanup -h -v --dry-run
Found 0 managed resource documents without models totaling 0B

Not removing 0 managed resources

Found 0 unreferenced resource documents totaling 0B

Not removing 0 dangling resources

Found 0 unreferenced blobstore files totaling 0B

Not removing 0 dangling files

Found 0 unreferenced blobstore chunks totaling 0B

Not removing 0 dangling chunks

total bytes: 0B
```
Passing --dry-run=false should remove anything.

## Documentation changes

None.

## Bug reference

None.

### CODE REVIEW

**Comment by jameinel** [comment]
We can look at 'what applications no longer refer to charms', but in my testing, it properly refcounted and removed content. eg 'juju upgrade-charm fat-charm --path ./fat-charm' would show disk consumption go up, and then drop a bit. It never went back to where it was, but it did recover some of the peak space. Repeated use also kept consuming a bit more disk space, but the blobstore.files and blobstore.chunks were being removed.

Agents is certainly a bit more obvious, we can look at the various model 'agent-version' configs. Though would we want to look at reported agent versions, or only *desired* agent versions, or do we flag those separately. (eg, there is a unit on agent-version 1.2.3, but it *should* be on 1.2.4, do we need to save the agent binary 1.2.3?)

**Comment by howbazaar** [comment]
I think the safest thing for the agent binaries is to save any versions that are referenced. At least to start.

**Comment by jameinel** [comment]
For those following along at home, this was run on one of our production controllers. Sorry for the Canonical private links.
Before:
https://juju-controller-reports.admin.canonical.com/jenkaas-is-mojo-ci/jenkins-0/juju-list-blobstore-2019-10-02T00.txt
https://juju-controller-reports.admin.canonical.com/jenkaas-is-mojo-ci/jenkins-0/juju-blobstore-cleanup-2019-10-02T00.txt

During cleanup:
https://juju-controller-reports.admin.canonical.com/jenkaas-is-mojo-ci/jenkins-0/juju-blobstore-cleanup-2019-10-02T00-2.txt

After cleanup:
https://juju-controller-reports.admin.canonical.com/jenkaas-is-mojo-ci/jenkins-0/juju-list-blobstore-2019-10-02T00-2.txt
https://juju-controller-reports.admin.canonical.com/jenkaas-is-mojo-ci/jenkins-0/juju-blobstore-cleanup-2019-10-02T00-3.txt

If there are ideas about how the output could be tightened up, made more manageable, I'm interested to hear them.

**Comment by jameinel** [comment]
!!build!!

flaky test:
https://bugs.launchpad.net/juju/+bug/1775737
https://jenkins.juju.canonical.com/job/github-make-check-juju/1296/testReport/junit/github/com_juju_juju_worker_raft/TestPackage/
FAIL: worker_test.go:281: WorkerSuite.TestNoLeaderTimeout
    c.Errorf("timed out waiting for worker to stop")
... Error: timed out waiting for worker to stop

File: scripts/juju-list-blobstore/main.go [hunk 1]
```diff
@@ -0,0 +1,1086 @@
...
+		unrefBytes += m.unreferencedBytes
+		totalBytes += m.totalBytes
+		refCharmBytes += m.appReferencedCharmBytes
+		unitrefCharmBytes += m.unitReferencedCharmBytes
+		unrefCharmBytes += m.unreferencedCharmBytes
+		refResourceBytes += m.referencedResourceBytes
+		unrefMiscBytes += m.unreferencedMiscBytes
+	}
+	fmt.Fprintf(os.Stdout, "\nModel Summary\n")
+	fmt.Fprintf(os.Stdout, "  app referenced charm bytes: %s\n", lengthToSize(refCharmBytes))
```
[the remark this answers is not in the document]
Comment (jameinel) [reply scripts/juju-list-blobstore/main.go]: only if you specify -h (human) but its a fair point.

File: scripts/juju-list-blobstore/main.go [hunk 2]
```diff
@@ -0,0 +1,1086 @@
...
+	fmt.Fprintf(os.Stdout, "  unit referenced charm bytes: %s\n", lengthToSize(unitrefCharmBytes))
```
[the remark this answers is not in the document]
Comment (jameinel) [reply scripts/juju-list-blobstore/main.go]: sure

File: scripts/juju-list-blobstore/main.go [hunk 3]
```diff
@@ -0,0 +1,1086 @@
...
+	if len(missingResources) == 0 {
+		fmt.Fprint(os.Stdout, "\nNo Unknown Resources\n")
+		return
+	}
+	resourceRefs := make(map[string][]managedResourceDoc, len(missingResources))
+	// Note, there isn't an index on resourceid, otherwise we'd just do a reverse lookup
+	var manageDoc managedResourceDoc
+	managedRefs := b.managedResources.Find(bson.M{"resourceid": bson.M{"$in": missingIds}}).Iter()
+	for managedRefs.Next(&manageDoc) {
+		// if _, missing := missingResources[manageDoc.ResourceId]; !missing {
```
[the remark this answers is not in the document]
Comment (jameinel) [reply scripts/juju-list-blobstore/main.go]: :+1:

File: scripts/juju-list-blobstore/main.go [hunk 4]
```diff
@@ -0,0 +1,1086 @@
...
+	}
+}
+
+func (b *BlobStoreChecker) readAgentBinaries() {
+	toolsStorage, err := b.system.ToolsStorage()
+	checkErr("tools storage", err)
+	defer toolsStorage.Close()
+
+	modelUUID := b.system.ModelUUID()
+	// managedStore := blobstore.NewManagedStorage(jujuDB, blobstore.NewGridFS("blobstore", "blobstore", session))
```
[the remark this answers is not in the document]
Comment (jameinel) [reply scripts/juju-list-blobstore/main.go]: 👍

**Comment by jameinel** [comment]
$$merge$$

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**